tests/math/test_functions.py

```diff
@@ -1100,6 +1100,117 @@ def test_unknown_interface(self):
             fn.requires_grad(type("hello", tuple(), {})())
 
 
+class TestInBackprop:
+    """Tests for the in_backprop function"""
+
+    @pytest.mark.slow
+    def test_jax(self):
+        """JAX DeviceArrays differentiability depends on the argnums argument"""
+        res = None
+
+        def cost_fn(t, s):
+            nonlocal res
+            res = [fn.in_backprop(t), fn.in_backprop(s)]
+            return jnp.sum(t * s)
+
+        t = jnp.array([1.0, 2.0, 3.0])
+        s = jnp.array([-2.0, -3.0, -4.0])
+
+        jax.grad(cost_fn, argnums=0)(t, s)
+        assert res == [True, False]
+
+        jax.grad(cost_fn, argnums=1)(t, s)
+        assert res == [False, True]
+
+        jax.grad(cost_fn, argnums=[0, 1])(t, s)
+        assert res == [True, True]
+
+    def test_autograd_backwards(self):
+        """Autograd trainability corresponds to the requires_grad attribute during the backwards pass."""
+        res = None
+
+        def cost_fn(t, s):
+            nonlocal res
+            res = [fn.in_backprop(t), fn.in_backprop(s)]
+            return np.sum(t * s)
+
+        t = np.array([1.0, 2.0, 3.0])
+        s = np.array([-2.0, -3.0, -4.0])
+
+        qml.grad(cost_fn)(t, s)
+        assert res == [True, True]
+
+        t.requires_grad = False
+        qml.grad(cost_fn)(t, s)
+        assert res == [False, True]
+
+        t.requires_grad = True
+        s.requires_grad = False
+        qml.grad(cost_fn)(t, s)
+        assert res == [True, False]
+
+        t.requires_grad = False
+        s.requires_grad = False
+        with pytest.warns(UserWarning, match="Attempted to differentiate a function with no"):
+            qml.grad(cost_fn)(t, s)
+        assert res == [False, False]
+
+    def test_tf(self):
+        """TensorFlow tensors will True *if* they are being watched by a gradient tape"""
+        t1 = tf.Variable([1.0, 2.0])
+        t2 = tf.constant([1.0, 2.0])
+        assert not fn.in_backprop(t1)
+        assert not fn.in_backprop(t2)
+
+        with tf.GradientTape():
+            # variables are automatically watched within a context,
+            # but constants are not
+            assert fn.in_backprop(t1)
+            assert not fn.in_backprop(t2)

```

Comment:
I'm curious what happens when you set `trainable=False` when initializing the `tf.Variable`. Do you know if `in_backprop` will return True or False in this scenario?

Reply:
I tried it it is working as intended 👍